controllers/machinehealthcheck_controller_test.go

```diff
@@ -1460,3 +1382,232 @@ func ownerReferenceForCluster(ctx context.Context, c *clusterv1.Cluster) metav1.
+	return clusterv1.Machine{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: clusterv1.GroupVersion.String(),
+			Kind:       "Machine",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "test-mhc-machine-",
+			Namespace:    mhc.Namespace,
+			Labels:       mhc.Spec.Selector.MatchLabels,
+		},
+		Spec: clusterv1.MachineSpec{
+			ClusterName: cluster.Name,
+			Bootstrap: clusterv1.Bootstrap{
+				DataSecretName: pointer.StringPtr("data-secret-name"),
+			},
+		},
+		Status: clusterv1.MachineStatus{
+			InfrastructureReady: true,
+			BootstrapReady:      true,
+			Phase:               string(clusterv1.MachinePhaseRunning),
+			ObservedGeneration:  1,
+		},
+	}
+}
+
+func newNode() corev1.Node {
+	node := corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "test-mhc-node-",
+		},
+		Status: corev1.NodeStatus{
+			Conditions: []corev1.NodeCondition{
+				{Type: corev1.NodeReady, Status: corev1.ConditionTrue},
+			},
+		},
+	}
+	return node
+}
+
+func setNodeUnhealthy(node *corev1.Node) {
+	node.Status.Conditions[0] = corev1.NodeCondition{
+		Type:               corev1.NodeReady,
+		Status:             corev1.ConditionUnknown,
+		LastTransitionTime: metav1.NewTime(time.Now().Add(-10 * time.Minute)),
+	}
+}
+
+func newInfraMachine(machine *clusterv1.Machine) (*unstructured.Unstructured, string) {
+	providerID := fmt.Sprintf("test:////%v", uuid.NewUUID())
+	return &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "infrastructure.cluster.x-k8s.io/v1alpha3",
+			"kind":       "InfrastructureMachine",
+			"metadata": map[string]interface{}{
+				"generateName": "test-mhc-machine-infra-",
+				"namespace":    machine.Namespace,
+			},
+			"spec": map[string]interface{}{
+				"providerID": providerID,
+			},
+		},
+	}, providerID
+}
+
+func fakeNodesMachines(
+	g *WithT,
+	n int,
+	healthy bool,
+	assignNodeRef bool,
+	mhc *clusterv1.MachineHealthCheck,
+	c *clusterv1.Cluster,
+) ([]*corev1.Node, []*clusterv1.Machine, func()) {
+
+	var (
+		nodes    []*corev1.Node
+		machines []*clusterv1.Machine
+	)
+
+	for i := 0; i < n; i++ {
+		machine := newMachine(mhc, c)
+		infraMachine, providerID := newInfraMachine(&machine)
+		node := newNode()
+		if !healthy {
+			setNodeUnhealthy(&node)
+		}
+		g.Expect(testEnv.Create(ctx, infraMachine)).To(Succeed())
+		fmt.Printf("inframachine created: %s\n", infraMachine.GetName())
+		machine.Spec.InfrastructureRef = corev1.ObjectReference{
+			APIVersion: infraMachine.GetAPIVersion(),
+			Kind:       infraMachine.GetKind(),
+			Name:       infraMachine.GetName(),
+		}
+		// TODO: Add comment on what's happening here. I think we are making
+		// the InfraMachine look like its healthy and ready.
+		infraMachinePatch := client.MergeFrom(infraMachine.DeepCopy())
+		g.Expect(unstructured.SetNestedField(infraMachine.Object, true, "status", "ready")).To(Succeed())
+		g.Expect(testEnv.Status().Patch(ctx, infraMachine, infraMachinePatch)).To(Succeed())
```

Comment:
Might be I'm missing something, but I find strange to create objects and then patch them (here and down in the func).
Should we instead create the objects, set fields according to the flags, and then create at the end when the objects are fully configured?

Reply:
TBH I didn't refactor this func and left the original behavior because I was unsure as to why we were doing some of these patches. I had added a bunch of comments asking why we were doing these things as well. 😄 

I can try refactoring this function to see if it can be improved upon further.

Reply:
ok, good catch.
AFAIK no, you can't set the status during create (status is to be set by the controllers).